# Predictive Analysis using scikit-learn
Name: Sam Sikder  
Course: IS 362  

## Objective
The goal of this project is to determine which feature (odor vs another column) better predicts whether a mushroom is poisonous.

## Introduction
In this project, I use machine learning with scikit-learn to analyze the mushroom dataset and predict whether a mushroom is poisonous or edible. 

The goal is to compare two predictor variables: **odor** and **gill-color**, and determine which one is more effective at predicting toxicity. 

This is important because identifying poisonous mushrooms accurately can prevent serious health risks. By applying classification models, we can evaluate how well different features contribute to prediction accuracy.

In [2]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score

In [3]:
url = "https://archive.ics.uci.edu/ml/machine-learning-databases/mushroom/agaricus-lepiota.data"

columns = [
    "class","cap-shape","cap-surface","cap-color","bruises","odor",
    "gill-attachment","gill-spacing","gill-size","gill-color",
    "stalk-shape","stalk-root","stalk-surface-above-ring",
    "stalk-surface-below-ring","stalk-color-above-ring",
    "stalk-color-below-ring","veil-type","veil-color",
    "ring-number","ring-type","spore-print-color",
    "population","habitat"
]

df = pd.read_csv(url, names=columns)

df.head()

,class,cap-shape,cap-surface,cap-color,bruises,odor,gill-attachment,gill-spacing,gill-size,gill-color,...,stalk-surface-below-ring,stalk-color-above-ring,stalk-color-below-ring,veil-type,veil-color,ring-number,ring-type,spore-print-color,population,habitat
0,p,x,s,n,t,p,f,c,n,k,...,s,w,w,p,w,o,p,k,s,u
1,e,x,s,y,t,a,f,c,b,k,...,s,w,w,p,w,o,p,n,n,g
2,e,b,s,w,t,l,f,c,b,n,...,s,w,w,p,w,o,p,n,n,m
3,p,x,y,w,t,p,f,c,n,n,...,s,w,w,p,w,o,p,k,s,u
4,e,x,s,g,f,n,f,w,b,k,...,s,w,w,p,w,o,e,n,a,g


## Data Loading

The dataset is loaded from the UCI Machine Learning Repository and contains categorical features describing different characteristics of mushrooms.

Each row represents a mushroom, and the **target variable (`class`)** indicates whether it is poisonous (`p`) or edible (`e`). 

Before building a predictive model, it is necessary to explore and prepare the dataset for analysis.

In [5]:
df_selected = df[["class", "odor", "gill-color"]]
df_selected.head()

,class,odor,gill-color
0,p,p,k
1,e,a,k
2,e,l,n
3,p,p,n
4,e,n,k


In [6]:
df_encoded = pd.get_dummies(df_selected, drop_first=True)
df_encoded.head()

,class_p,odor_c,odor_f,odor_l,odor_m,odor_n,odor_p,odor_s,odor_y,gill-color_e,gill-color_g,gill-color_h,gill-color_k,gill-color_n,gill-color_o,gill-color_p,gill-color_r,gill-color_u,gill-color_w,gill-color_y
0,True,False,False,False,False,False,True,False,False,False,False,False,True,False,False,False,False,False,False,False
1,False,False,False,False,False,False,False,False,False,False,False,False,True,False,False,False,False,False,False,False
2,False,False,False,True,False,False,False,False,False,False,False,False,False,True,False,False,False,False,False,False
3,True,False,False,False,False,False,True,False,False,False,False,False,False,True,False,False,False,False,False,False
4,False,False,False,False,False,True,False,False,False,False,False,False,True,False,False,False,False,False,False,False


## Data Preprocessing

The selected features are categorical, meaning they contain text values rather than numerical values.

Machine learning models require numerical input, so I applied **one-hot encoding** using `pd.get_dummies()` to transform each category into binary columns.

This process creates separate columns for each possible value of odor and gill-color, allowing the model to interpret these features correctly.

The parameter `drop_first=True` was used to avoid redundancy and reduce multicollinearity by removing one category from each feature.

## Model Building

A **Decision Tree Classifier** was used to build the prediction model. This model is suitable for categorical data and can identify patterns in how features relate to the target variable.

The dataset was split into training and testing sets:
- **Training set**: used to train the model
- **Testing set**: used to evaluate performance

Accuracy was used as the evaluation metric to measure how well the model predicts whether a mushroom is poisonous.

In [9]:
X = df_encoded.drop("class_p", axis=1)  # features
y = df_encoded["class_p"]               # target (poisonous or not)

In [10]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

In [11]:
model = DecisionTreeClassifier()
model.fit(X_train, y_train)

y_pred = model.predict(X_test)

accuracy = accuracy_score(y_test, y_pred)
print("Accuracy (odor + gill-color):", accuracy)

Accuracy (odor + gill-color): 0.9876948318293683


## Model Using Odor Only

In this step, I trained a model using only the odor-related features.

This allows me to evaluate how strong odor is as a single predictor of mushroom toxicity. If the accuracy is high, it suggests that odor alone provides significant predictive power.

In [13]:
X_odor = df_encoded.filter(like="odor")
X_train, X_test, y_train, y_test = train_test_split(X_odor, y, test_size=0.3, random_state=42)

model = DecisionTreeClassifier()
model.fit(X_train, y_train)

y_pred = model.predict(X_test)

print("Accuracy (odor only):", accuracy_score(y_test, y_pred))

Accuracy (odor only): 0.9840032813781788


## Model Using Gill-Color Only

In this step, I trained a model using only gill-color features.

This helps determine whether visual characteristics like gill color are as effective as odor in predicting whether a mushroom is poisonous.

In [15]:
X_gill = df_encoded.filter(like="gill-color")
X_train, X_test, y_train, y_test = train_test_split(X_gill, y, test_size=0.3, random_state=42)

model = DecisionTreeClassifier()
model.fit(X_train, y_train)

y_pred = model.predict(X_test)

print("Accuracy (gill-color only):", accuracy_score(y_test, y_pred))

Accuracy (gill-color only): 0.8018867924528302


## Results Comparison

The accuracy scores from each model were compared to determine which feature performed better.

A higher accuracy indicates that the feature is more reliable for predicting whether a mushroom is poisonous.

The odor-only model had an accuracy of about 98.4%, while the gill-color-only model had an accuracy of about 80.2%. The model using both odor and gill-color had the highest accuracy at about 98.8%.

## Conclusion

Based on the results of the predictive models, odor was the stronger predictor of whether a mushroom is poisonous compared to gill-color. The model using odor produced a higher accuracy score, indicating a stronger relationship between this feature and the target variable. This suggests that odor is a highly informative characteristic when determining mushroom toxicity.

For further analysis, I would test additional features such as spore-print-color, habitat, and bruises to see if they improve prediction accuracy. Additionally, combining multiple features into a single model could provide more accurate and reliable results compared to using only one predictor at a time.